### ISSUE TITLE
Custom DataType DB Objects

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#1109]
I'm looking to experiment with a custom DataType to handle accessing of a structures virtual members. I can create a class ending in DataType so the classloader loads it and I get the intended behavior until it gets added to a DataTypeManager. I haven't found anything in the API, but is there a way I can register a callback for a corresponding DB class for when my DataType is added to the DataTypeManager? I would like it to be able to work in a plugin. 

I think I could extend an existing DataTypeManager to get it to create the DB object but I'm uncertain if the correct DB object will appear in the listing.

**Comment by ghidra1** [comment]
This can get rather confusing as to what technique to use.   New DB implementations should not be  considered an option, although BuiltIn types may. If the result is an in-memory structure, you can consider either a FactoryDataType or an analyzer, both result in applied Structure datatypes applied to memory, the resulting in-memory structures would be recognized by the Decompiler. 
 Other approaches have the following drawbacks:
1. Decompiler will not know how to handle adhoc BuiltIn types
2. If new BuiltIn type is not included within a Ghidra distribution it will be difficult to share project files
3. If type is Dynamic it may only be applied to initialized memory or in pointer form with Function variables/parameters.
I am sure I omitted other considerations.  You may want to look at the approach adopted by the MicrosoftCodeAnalyzer module for applying RTTI data structures.

**Comment by astrelsky** [comment]
@ghidra1 I actually did mirror the MicrosoftCodeAnalyzer module for the creation of gcc RTTI data structures in my [C++ Class Analyzer](https://github.com/astrelsky/Ghidra-Cpp-Class-Analyzer/tree/master/src/main/java/ghidra/app/cmd/data/rtti/gcc/typeinfo)

This is for the same thing I'm trying to do with #1016 . I want to be able to tell the decompiler that when it accesses a structures _vptr field that is is pulling a ptrdiff_t from the offset_to_virtual_base array. My hope is that it will take the constant in the array, apply it and resolve the correct field. From the perspective of the decompiler, the offset of the virtual member at runtime doesn't actually matter. For decompiling purposes it is only necessary to know to access the offset of the virtual member in the structure of the this parameter.

I was at one point able to do it with a StructureDataType which I tested in python. It just didn't work once added to the DataTypeManager because the manager would create a StructureDB which wouldn't replicate the behavior. I scrapped the idea and came back to it this morning wondering if it was actually possible.

**Comment by ghidra1** [comment]
Did you isolate why the StructureDataType and StructureDB did not behave the same.  This sounds fishy and is worth identifying the cause, particularly since it would be a much cleaner solution if it works.  These two implementation are suppose to act the same given they are associated with the same DataTypeManager / DataOrganization.

**Comment by astrelsky** [comment]
> Did you isolate why the StructureDataType and StructureDB did not behave the same. This sounds fishy and is worth identifying the cause, particularly since it would be a much cleaner solution if it works. These two implementation are suppose to act the same given they are associated with the same DataTypeManager / DataOrganization.

I had created a DataType which extended StructureDataType. This allowed me to override the behavior when accessed with a negative index. When adding it to a DataTypeManager it sees it is a Structure instance and creates a StructureDB instance for it which doesn't contain the desired behavior.

**Comment by ghidra1** [comment]
Ooo... I guess you found out the hard way you should not attempt to extend any datatype which is not a BuiltInDataType since other types must resolve to a corresponding DataTypeDB.  Any new type you create should extend BuiltInDataType, although as mentioned the decompiler will not know how to cope with such types.  Adding new DB DataTypes is not something we recommend due to the very large impact it would have.

We are aware of the case you are trying to solve and others like it, although I don't know the timeline for a solution.  I would be curious to see your custom Structure implementation which seemed to solve your case.

**Comment by astrelsky** [comment]
It was just a prototype. It may work in theory but it depends on how the decompiler fetches the component. All I did was override getComponentAt and a few other methods to accept a negative input. Also virtual inheritance is implementation defined so how the data is laid out is compiler specific. Visual Studio has the vbase array directly after the vftable while gcc has it directly before it.

Now that I'm thinking about it again, having a "pointer to member" datatype might work. That is assuming the constants will get resolved if I set the value of the varnode to the correct address.

I guess another possibility with the way things currently are would be to make the structure at the start of the base array. I would then have to increment the value of the varnode to correctly offset it.

Probably going to have to go back to the drawing board. If I need to do testing with a DataTypeDB I'll likely patch in a way of creating a callback for faster testing. I am hoping I can get away with just manipulating the varnodes and references.

Thank you for your time.
